source/request_source/request_options_list_plugin_impl.h

```diff
@@ -45,40 +44,69 @@ class RequestOptionsListRequestSource : public RequestSource {
-  // or could not be found.
+  // This implementation is not thread safe. There is a race condition because only the first call
+  // to createRequestSourcePlugin will load the file from memory and subsequent calls just make a
+  // copy of the options_list that was already loaded. The FileBasedOptionsListRequestSourceFactory
+  // will not work with multiple different files for this reason. This method will also error if the
+  // file can not be loaded correctly, e.g. the file is too large or could not be found.
   RequestSourcePtr createRequestSourcePlugin(const Envoy::Protobuf::Message& message,
                                              Envoy::Api::Api& api,
                                              Envoy::Http::RequestHeaderMapPtr header) override;
 
 private:
   Envoy::Thread::MutexBasicLockable file_lock_;
-  nighthawk::client::RequestOptionsList options_list_;
+  absl::optional<nighthawk::client::RequestOptionsList> options_list_;
 };
 
 // This factory will be activated through RequestSourceFactory in factories.h
-DECLARE_FACTORY(OptionsListFromFileRequestSourceFactory);
+DECLARE_FACTORY(FileBasedOptionsListRequestSourceFactory);
+
+// Factory that creates a OptionsListRequestSource from a InLineOptionsListRequestSourceConfig
+// proto. Registered as an Envoy plugin. Implementation of RequestSourceConfigFactory which produces
+// a RequestSource that keeps an RequestOptionsList in memory, and loads it with the RequestOptions
+// passed to it from the config. All plugins configuration are specified in the
+// request_source_plugin.proto. Usage: assume you are passed an appropriate Any type object called
+// config, an Api object called api, and a default header called header. auto& config_factory =
+//     Envoy::Config::Utility::getAndCheckFactoryByName<RequestSourcePluginConfigFactory>(
+//         "nighthawk.in-line-options-list-request-source-plugin");
+// RequestSourcePtr plugin =
+//     config_factory.createRequestSourcePlugin(config, std::move(api), std::move(header));
+
+class InLineOptionsListRequestSourceFactory : public virtual RequestSourcePluginConfigFactory {
+public:
+  std::string name() const override;
+  Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+
+  // This implementation is not thread safe. There is a race condition because only the first call
+  // to createRequestSourcePlugin will load the options list into memory and subsequent calls just
+  // make a copy of the options_list that was already loaded.
+  RequestSourcePtr createRequestSourcePlugin(const Envoy::Protobuf::Message& message,
+                                             Envoy::Api::Api& api,
+                                             Envoy::Http::RequestHeaderMapPtr header) override;
+
+private:
+  Envoy::Thread::MutexBasicLockable options_list_lock_;
+  absl::optional<nighthawk::client::RequestOptionsList> options_list_;
```

Comment:
Does this field need to be optional? When would we create an instance of InLineOptionsListRequestSourceFactory without an options_list_?

Reply:
I made this optional to indicate the state before and after having saved the options_list_.